tuda_workspace_scripts/discovery.py

```diff
@@ -0,0 +1,126 @@
+from .workspace import get_workspace_root
+from robots import DiscoveryServer, load_robots
+from tuda_workspace_scripts import print_warn
+from ros2cli.node.daemon import is_daemon_running, spawn_daemon, shutdown_daemon
+import os
+import xml.etree.ElementTree as ET
+
+
+# dedicated place for temporary files
+tmp_directory = "/tmp/tuda_wss"
+discovery_xml_path = os.path.join(tmp_directory, "discovery_client.xml")
+super_client_xml_path = os.path.join(tmp_directory, "super_client.xml")
+
+def create_discovery_xml(discovery_server_names: list[str]):
+  if not os.path.exists(tmp_directory):
+    os.mkdir(tmp_directory)
+
+  discovery_servers = []
+  robots = load_robots()
+  for name in discovery_server_names:
+    if name == "local_server":
+      discovery_servers.append(DiscoveryServer("127.0.0.1", 11811, "44.53.00.5f.45.50.52.4f.53.49.4d.41"))
+    elif name == "all":
+      for __, robot in robots.items():
+        discovery_servers.extend(robot.discovery_servers)
+      # This is currently unclean because there should be a better toggle/standard
+      discovery_servers.append(DiscoveryServer("127.0.0.1", 11811, "44.53.00.5f.45.50.52.4f.53.49.4d.41"))
+      # No servers can be added anymore, so all other arguments are ignored
+      break
+    else:
+      # Only adding if there is exactly one robot, else the robots xml is probably misformed ore the robot is missing.
+      # check for robots with correct name
+      filtered_robots = []
+      for robot_name, robot_data in robots.items():
+        if robot_name == name:
+          filtered_robots.append(robot_data)
+      if len(filtered_robots) == 1:
+        discovery_servers.extend(filtered_robots[0].discovery_servers)
+      elif len(filtered_robots) == 0:
+        print_warn(f"Couldn't find correct entry for {name} in robot configs. Please check if your selected robot is available.")
+      else:
+        print_warn(f"Found multiple robot entries for {name} in robot configs. Your configs seem to be misformed.")
+  
+  root = _create_dds_xml(discovery_servers)
+
+  # pretty printing
+  ET.indent(root, '    ')
+
+  # Create temporary file so it is removed on reboot
+  tree = ET.ElementTree(root)
+  tree.write(discovery_xml_path, encoding="utf-8", xml_declaration=True)
+
+  # Create Super client xml
+  protocol = root.findall(f".//discoveryProtocol")[0]
+  protocol.text = "SUPER_CLIENT"
+  tree.write(super_client_xml_path, encoding="utf-8", xml_declaration=True)
+
+def _create_dds_xml(discovery_servers: list[DiscoveryServer]) -> ET.Element:
```

Comment:
Currently this is only FastRTPS, can we abstract this more in some way so we can more easily add support for other middleware?

Reply:
From what I have read, it seems that the fastdds server is for FastRtps only. It seems no to be directly supported by other middleware. It could be possible, but it seems much more complicated.

Reply:
Yes, but other middlewares have analogues for the discovery server. E.g. zenoh has routers.